pkg/worker/redis.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	cmd := w.client.LPush(ctx, w.queueName, buffer.Bytes())
+	if cmd.Err() != nil {
+		logger.Error().Err(err).Msg("Unable to push job into Redis")
+		return fmt.Errorf("unable to push job into Redis: %w", cmd.Err())
+	}
+
+	result, err := cmd.Result()
+	if err != nil {
+		return fmt.Errorf("unable to process result: %w", err)
+	}
+	logger.Info().Int64("job_result", result).Msg("Pushed job successfully")
+	return nil
+}
+
+func (w *RedisWorker) Stop(ctx context.Context) {
+	logger := ctxval.Logger(ctx)
+	close(w.closeCh)
+	logger.Info().Msg("Waiting for all workers to finish")
+	w.loopWG.Wait()
+	w.handleWG.Wait()
+	logger.Info().Msg("Done waiting for all workers to finish")
+}
+
+func (w *RedisWorker) DequeueLoop(ctx context.Context) {
+	logger := ctxval.Logger(ctx)
+	pollers := math.Min(w.pollMax, runtime.NumCPU()+1)
+	logger.Info().Msgf("Starting %d polling goroutines", pollers)
+	for i := 1; i <= pollers; i++ {
+		w.loopWG.Add(1)
+		go w.dequeueLoop(ctx)
+	}
+}
+
+//nolint:gosec
+func (w *RedisWorker) dequeueLoop(ctx context.Context) {
+	defer w.loopWG.Done()
+	logger := ctxval.Logger(ctx)
+
+	// randomize polling intervals for a better spread (random delay up to one read timeout)
+	randMs := rand.Int63n(w.pollInterval.Milliseconds())
+	logger.Info().Msgf("Worker start sleep (ms): %d", randMs)
+	time.Sleep(time.Duration(randMs) * time.Millisecond)
+
+	for {
+		select {
+		case <-w.closeCh:
+			logger.Info().Msg("Shutting down a Redis poller (stop)")
+			return
+		case <-ctx.Done():
+			logger.Info().Msg("Shutting down a Redis poller (cancel)")
+			return
+		default:
+			w.fetchJob(ctx)
+		}
+	}
+}
+
+func (w *RedisWorker) fetchJob(ctx context.Context) {
+	res, err := w.client.BLPop(ctx, w.pollInterval, w.queueName).Result()
+
+	if errors.Is(err, redis.Nil) {
+		// timeout occurred
+	} else if err != nil {
+		logger := ctxval.Logger(ctx)
+		logger.Error().Err(err).Msg("Error consuming from Redis queue")
+	} else {
+		var job Job
+		dec := gob.NewDecoder(strings.NewReader(res[1]))
+		err = dec.Decode(&job)
+		logger := loggerWithJob(ctx, &job)
+		if err != nil {
+			logger.Error().Err(err).Msg("Unable to unmarshal job payload, skipping")
+		}
+
+		w.handleWG.Add(1)
+		atomic.AddInt64(&w.inFlight, 1)
+		go w.processJob(ctx, &job)
+	}
+}
+
+func (w *RedisWorker) processJob(ctx context.Context, job *Job) {
+	defer w.handleWG.Done()
+	defer atomic.AddInt64(&w.inFlight, -1)
+	logger := loggerWithJob(ctx, job)
+
+	logger.Info().Msg("Dequeued job from Redis")
+	if h, ok := w.handlers[job.Type]; ok {
+		h(ctx, job)
+	} else {
+		// handler not found
+		ctxval.Logger(ctx).Warn().Msgf("Redis worker handler not found for job type: %s", job.Type)
+	}
+}
+
+func (w *RedisWorker) Stats(ctx context.Context) (Stats, error) {
+	count, err := w.client.LLen(ctx, w.queueName).Result()
+	if err != nil {
+		return Stats{}, fmt.Errorf("unable to get queue len: %w", err)
+	}
+
+	return Stats{
+		EnqueuedJobs: uint64(count),
+		InFlight:     w.inFlight,
+	}, nil
+}
```

Comment:
And this is the "heart" of the stage/prod environments - Redis implementation, uses atomic list operation for the queue, it has a single list for all tasks, each task has a type (string, like a "channel"), UUID (random id for logging, no other purpose) and args which is a generic struct.

It uses gob encoder from the Go stdlib to encode/decode jobs which allows to do type assertion of `any` interface in a safe way (e.g. when a new field is added gob will safely ignore it - leave it blank).

Graceful shutdown is implemented via wait groups, the `Stop` function waits for all workers to finish polling as well as for all goroutines to finish processing of already started jobs.